File: 909-snakes-and-ladders/solution.py
from collections import deque


class Solution:
  @staticmethod
  def get_board_val(num: int, board: list[list[int]]) -> int:
    num = num - 1  # change num from being 1-indexed to 0-indexed
    n = len(board)
    row_num = n - 1 - int(num / n)
    col_num = num % n if (n - row_num + 1) % 2 == 0 else n - 1 - (num % n)
    return board[row_num][col_num]

  def snakesAndLadders(self, board: list[list[int]]) -> int:
    target = len(board) ** 2
    visited: set[int] = set()
    work_deque: deque[tuple[int, int]] = deque([(1, 0)])
    while work_deque:
      num, step = work_deque.popleft()
      if num == target:
        return step

      if num in visited:
        continue

      visited.add(num)
      for i in range(1, 7):
        if num + i > target:
          break

        board_val = Solution.get_board_val(num + i, board)
        work_deque.append(
            (num + i if board_val == -1 else board_val, step + 1))

    return -1

File: 909-snakes-and-ladders/solution.cpp
#include <vector>
#include <cmath>
#include <unordered_set>
#include <deque>
#include <tuple>

class Solution
{
private:
  static int get_board_val(int num, const std::vector<std::vector<int>> &board)
  {
    num = num - 1;
    const int n = board.size();
    const int row_num = n - 1 - (num / n);
    const int col_num = (n - row_num + 1) % 2 == 0 ? num % n : n - 1 - (num % n);
    return board[row_num][col_num];
  }

public:
  int snakesAndLadders(std::vector<std::vector<int>> &board)
  {
    const int target = std::pow(board.size(), 2);
    std::unordered_set<int> visited;
    std::deque<std::tuple<int, int>> work_deque{{1, 0}};
    while (!work_deque.empty())
    {
      const auto num_step = work_deque.front();
      work_deque.pop_front();
      const int num = std::get<0>(num_step);
      const int step = std::get<1>(num_step);
      if (num == target)
      {
        return step;
      }

      if (visited.contains(num))
      {
        continue;
      }

      visited.emplace(num);
      for (int i = 1; i < 7; ++i)
      {
        if (num + i > target)
        {
          break;
        }

        const int board_val = Solution::get_board_val(num + i, board);
        work_deque.emplace_back(board_val == -1 ? num + i : board_val, step + 1);
      }
    }

    return -1;
  }
};